<h1 align="center" style="margin-bottom: 20px;">Machine Learning 2024-25</h1>
<h3 align="center" style="margin-top: 20px; margin-bottom: 20px;">To Grant or Not to Grant: Deciding on Compensation Benefits</h3>
<h5 align="center" style="margin-top: 20px; margin-bottom: 0px;">Cross-Validation for Models with Default Parameters</h5>

### Group 38 - Members:
- Ana Marta Azinheira  - 20240496@novaims.unl.pt
- Braulio Damba - 20240007@novaims.unl.pt
- Catarina Ribeirinha - 20240507@novaims.unl.pt
- Marco Galão  - r20201545@novaims.unl.pt
- Rodrigo Sardinha - 20211627@novaims.unl.pt

# Table of Contents

* [1. Imports](#imports)
  * [1.1. Import Libraries](#import_libraries)
  * [1.2. Imports Datasets](#import_datasets)

* [2. Cross-Validation for Models with Default Parameters](#cv_default_parameters)

# 1. Imports <a id="imports"></a>

## 1.1. Import the Libraries <a class="anchor" id="import_libraries"></a>


In [487]:
# Cross Validation Utils
from utils import *

# Visualization Settings
%config InlineBackend.figure_format = 'retina'
sns.set()

# Omit Warnings
import warnings
warnings.filterwarnings("ignore")

## 1.2. Import the Datasets <a class="anchor" id="import_datasets"></a>

In [488]:
# Define date columns to parse
date_cols = ["Accident Date", "Assembly Date", "C-2 Date", "C-3 Date", "First Hearing Date"]

In [489]:
df = pd.read_csv('df_nb2.csv', sep=',', parse_dates=date_cols, index_col='Claim Identifier')
df.head()

Accident Date  Age at Injury  Alternative Dispute Resolution  \
Claim Identifier                                                                
5393875             2019-12-30             31                             0.0   
5393091             2019-08-30             46                             0.0   
5393889             2019-12-06             40                             0.0   
5393887             2019-12-30             61                             0.0   
5393863             2019-12-26             67                             0.0   

                 Assembly Date  Attorney/Representative  Average Weekly Wage  \
Claim Identifier                                                               
5393875             2020-01-01                        0                 0.00   
5393091             2020-01-01                        1              1745.93   
5393889             2020-01-01                        0              1434.80   
5393887             2020-01-01                        0                  NaN   
5393863             2020-01-01                        0                 0.00   

                   C-2 Date   C-3 Date                       Carrier Name  \
Claim Identifier                                                            
5393875          2019-12-31        NaT    new hampshire insurance company   
5393091          2020-01-01 2020-01-01  zurich american insurance company   
5393889          2020-01-01        NaT     indemnity insurance company of   
5393887          2019-12-31        NaT               state insurance fund   
5393863          2019-12-31        NaT   indemnity insurance of n america   

                 Carrier Type  ...  \
Claim Identifier               ...   
5393875           1A. PRIVATE  ...   
5393091           1A. PRIVATE  ...   
5393889           1A. PRIVATE  ...   
5393887               2A. SIF  ...   
5393863           1A. PRIVATE  ...   

                                          Industry Code Description  \
Claim Identifier                                                      
5393875                                                RETAIL TRADE   
5393091                                                CONSTRUCTION   
5393889           ADMINISTRATIVE AND SUPPORT AND WASTE MANAGEMEN...   
5393887                           HEALTH CARE AND SOCIAL ASSISTANCE   
5393863                                                RETAIL TRADE   

                 Medical Fee Region  WCIO Cause of Injury Code  \
Claim Identifier                                                 
5393875                           I                       27.0   
5393091                           I                       97.0   
5393889                          II                       79.0   
5393887                          II                       16.0   
5393863                          IV                       31.0   

                 WCIO Cause of Injury Description WCIO Nature of Injury Code  \
Claim Identifier                                                               
5393875              FROM LIQUID OR GREASE SPILLS                       10.0   
5393091                         REPETITIVE MOTION                       49.0   
5393889            OBJECT BEING LIFTED OR HANDLED                        7.0   
5393887           HAND TOOL, UTENSIL; NOT POWERED                       43.0   
5393863                   FALL, SLIP OR TRIP, NOC                       10.0   

                  WCIO Nature of Injury Description WCIO Part Of Body Code  \
Claim Identifier                                                             
5393875                                   CONTUSION                   62.0   
5393091                              SPRAIN OR TEAR                   38.0   
5393889                                  CONCUSSION                   10.0   
5393887                                    PUNCTURE                   36.0   
5393863                                   CONTUSION                   38.0   

            

# 2. Cross-Validation for Models with Default Parameters <a id="cv_default_parameters"></a>

In [509]:
# Define default models
models_default_parameters = {
    "LogisticRegression": {},
    "RandomForest": {},
    "XGBoost": {},
    "SVM": {},
    "Bagging": {},
    "DecisionTree": {},
    "NeuralNetworks": {},
    "Adaboost": {},
    "GBM": {}
}

# Using default models to choose which ones to optimize
default_models = define_models(models_default_parameters)
show_results(default_models, df, feats_dict)

--------------------------------------------------------------------------------
CV Iteration 1
--------------------------------------------------------------------------------

Spearman Correlation - Number of Excluded Features: 5 / 20
Spearman Correlation - Excluded Features: {'Average Weekly Wage Category', 'Log Age at Injury', 'Medical Fee Region', 'County of Injury', 'Average Weekly Wage'}
Spearman Correlation - Highly Correlated Pairs: [('Average Weekly Wage Category', 'Average Weekly Wage'), ('Log Age at Injury', 'Age at Injury'), ('Age * Avg Weekly Wage', 'Average Weekly Wage'), ('Age * Avg Weekly Wage', 'Average Weekly Wage Category'), ('Average Weekly Wage by Medical Fee Region', 'County of Injury'), ('Average Weekly Wage by Medical Fee Region', 'Medical Fee Region')]

RFE - Number of Selected Features: 18 / 24
RFE - Selected Features: ['Age at Injury', 'Attorney/Representative', 'Average Weekly Wage', 'Carrier Type', 'County of Injury', 'COVID-19 Indicator', 'District Name',

F1 Val        F1 Train     F1 Val-Train  \
NeuralNetworks      0.380 +/- 0.04  0.611 +/- 0.06  -0.231 +/- 0.08   
Bagging             0.372 +/- 0.02  0.970 +/- 0.01  -0.599 +/- 0.03   
XGBoost             0.369 +/- 0.02  0.998 +/- 0.00  -0.629 +/- 0.03   
GBM                 0.362 +/- 0.02  0.941 +/- 0.02  -0.580 +/- 0.03   
LogisticRegression  0.352 +/- 0.05  0.440 +/- 0.03  -0.088 +/- 0.04   
DecisionTree        0.352 +/- 0.04  0.998 +/- 0.00  -0.646 +/- 0.04   
RandomForest        0.347 +/- 0.01  0.998 +/- 0.00  -0.651 +/- 0.02   
SVM                 0.317 +/- 0.01  0.433 +/- 0.03  -0.116 +/- 0.02   
Adaboost            0.259 +/- 0.06  0.298 +/- 0.09  -0.039 +/- 0.04   

                          Time (s)       Iterations Best Params  
NeuralNetworks      0.562 +/- 0.42    200.0 +/- 0.0        None  
Bagging             0.031 +/- 0.01      0.0 +/- 0.0        None  
XGBoost             0.294 +/- 0.12      0.0 +/- 0.0        None  
GBM                 0.594 +/- 0.07      0.0 +/- 0.0        None  
LogisticRegression  0.021 +/- 0.01     57.8 +/- 5.1        None  
DecisionTree        0.005 +/- 0.00      0.0 +/- 0.0        None  
RandomForest        0.154 +/- 0.02      0.0 +/- 0.0        None  
SVM                 0.026 +/- 0.01  204.0 +/- 252.8        None  
Adaboost            0.090 +/- 0.03      0.0 +/- 0.0        None

## Alterar as conclusões!!

**Best Default Models (or Worst Default Models): XGBoost**
- Logistic Regression: Low F1 scores (0.36 for both train and test), indicating poor performance.
- Random Forest: High train F1 score (1.0) and good test F1 score (0.72), indicating strong model performance but very likely overfitting.
- XGBoost: Balanced performance with good and very similar F1 scores for train and test (0.74 and 0.73, respectively), indicating robustness without overfitting.